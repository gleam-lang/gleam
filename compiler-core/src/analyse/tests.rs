use super::*;

#[test]
fn module_name_validation() {
    assert!(validate_module_name(&"dream".into()).is_ok());

    assert!(validate_module_name(&"gleam".into()).is_err());

    assert!(validate_module_name(&"gleam/ok".into()).is_ok());

    assert!(validate_module_name(&"ok/gleam".into()).is_ok());

    assert!(validate_module_name(&"external".into()).is_err());

    assert!(validate_module_name(&"type".into()).is_err());

    assert!(validate_module_name(&"pub".into()).is_err());

    assert!(validate_module_name(&"ok/external".into()).is_err());

    assert!(validate_module_name(&"ok/type".into()).is_err());

    assert!(validate_module_name(&"ok/pub".into()).is_err());
}
