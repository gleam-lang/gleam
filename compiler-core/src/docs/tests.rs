use std::{collections::HashSet, time::SystemTime};

use crate::{
    build::{Mode, NullTelemetry, PackageCompiler, StaleTracker, TargetCodegenConfiguration},
    config::{DocsPage, PackageConfig, Repository},
    docs::DocContext,
    io::{memory::InMemoryFileSystem, FileSystemWriter},
    paths::ProjectPaths,
    uid::UniqueIdGenerator,
    version::COMPILER_VERSION,
    warning::WarningEmitter,
};
use camino::Utf8PathBuf;
use ecow::EcoString;
use itertools::Itertools;

fn compile_with_markdown_pages(
    config: PackageConfig,
    modules: Vec<(&str, &str)>,
    markdown_pages: Vec<(&str, &str)>,
) -> EcoString {
    let fs = InMemoryFileSystem::new();
    for (name, src) in modules {
        fs.write(&Utf8PathBuf::from(format!("/src/{name}")), src)
            .unwrap();
    }

    // We're saving the pages under a different `InMemoryFileSystem` for these
    // tests so we don't have to juggle with borrows and lifetimes.
    // The package compiler is going to take ownership of `fs` but later
    // `generate_html` also needs a `FileSystemReader` to go and read the
    // markdown pages' content.
    let pages_fs = InMemoryFileSystem::new();
    for (title, src) in markdown_pages.iter() {
        pages_fs
            .write(&Utf8PathBuf::from(format!("{title}.md")), src)
            .unwrap();
    }

    let ids = UniqueIdGenerator::new();
    let mut type_manifests = im::HashMap::new();
    let mut defined_modules = im::HashMap::new();
    let warnings = WarningEmitter::null();
    let target = TargetCodegenConfiguration::Erlang { app_file: None };

    let root = Utf8PathBuf::from("/");
    let build = root.join("build");
    let lib = root.join("lib");
    let paths = ProjectPaths::new(root.clone());
    let mut compiler =
        PackageCompiler::new(&config, Mode::Dev, &root, &build, &lib, &target, ids, fs);
    compiler.write_entrypoint = false;
    compiler.write_metadata = false;
    compiler.compile_beam_bytecode = true;
    let mut modules = compiler
        .compile(
            &warnings,
            &mut type_manifests,
            &mut defined_modules,
            &mut StaleTracker::default(),
            &mut HashSet::new(),
            &NullTelemetry,
        )
        .unwrap();

    for module in &mut modules {
        module.attach_doc_and_module_comments();
    }

    let docs_pages = markdown_pages
        .into_iter()
        .map(|(title, _)| DocsPage {
            title: (*title).into(),
            path: format!("{title}.html"),
            source: format!("{title}.md").into(),
        })
        .collect_vec();

    super::generate_html(
        &paths,
        &config,
        &modules,
        &docs_pages,
        pages_fs,
        SystemTime::UNIX_EPOCH,
        DocContext::HexPublish,
    )
    .into_iter()
    .filter(|file| file.path.extension() == Some("html"))
    .sorted_by(|a, b| a.path.cmp(&b.path))
    .flat_map(|file| {
        Some(format!(
            "//// {}\n\n{}\n\n",
            file.path.as_str(),
            file.content
                .text()?
                .replace(COMPILER_VERSION, "GLEAM_VERSION_HERE")
        ))
    })
    .collect::<String>()
    .chars()
    .collect()
}

pub fn compile(config: PackageConfig, modules: Vec<(&str, &str)>) -> EcoString {
    compile_with_markdown_pages(config, modules, vec![])
}

#[test]
fn hello_docs() {
    let config = PackageConfig::default();
    let modules = vec![(
        "app.gleam",
        r#"
/// Here is some documentation
pub fn one() {
  1
}
"#,
    )];
    insta::assert_snapshot!(compile(config, modules));
}

// https://github.com/gleam-lang/gleam/issues/2347
#[test]
fn tables() {
    let config = PackageConfig::default();
    let modules = vec![(
        "app.gleam",
        r#"
/// | heading 1    | heading 2    |
/// |--------------|--------------|
/// | row 1 cell 1 | row 1 cell 2 |
/// | row 2 cell 1 | row 2 cell 2 |
///
pub fn one() {
  1
}
"#,
    )];
    insta::assert_snapshot!(compile(config, modules));
}

// https://github.com/gleam-lang/gleam/issues/2202
#[test]
fn long_function_wrapping() {
    let config = PackageConfig::default();
    let modules = vec![(
        "app.gleam",
        r#"
pub type Option(t) {
  Some(t)
  None
}

/// Returns the first value if it is `Some`, otherwise evaluates the given
/// function for a fallback value.
///
pub fn lazy_or(first: Option(a), second: fn() -> Option(a)) -> Option(a) {
  case first {
    Some(_) -> first
    None -> second()
  }
}
"#,
    )];

    insta::assert_snapshot!(compile(config, modules));
}

#[test]
fn internal_definitions_are_not_included() {
    let config = PackageConfig::default();
    let modules = vec![(
        "app.gleam",
        r#"
@internal
pub const wibble = 1

@internal
pub type Wibble = Int

@internal
pub type Wobble { Wobble }

@internal
pub fn one() { 1 }
"#,
    )];
    insta::assert_snapshot!(compile(config, modules));
}

// https://github.com/gleam-lang/gleam/issues/2561
#[test]
fn discarded_arguments_are_not_shown() {
    let config = PackageConfig::default();
    let modules = vec![("app.gleam", "pub fn discard(_discarded: a) -> Int { 1 }")];
    insta::assert_snapshot!(compile(config, modules));
}

// https://github.com/gleam-lang/gleam/issues/2631
#[test]
fn docs_of_a_type_constructor_are_not_used_by_the_following_function() {
    let config = PackageConfig::default();
    let modules = vec![(
        "app.gleam",
        r#"
pub type Wibble {
  Wobble(
    /// Documentation!!
    wabble: Int,
  )
}

pub fn main() { todo }
"#,
    )];
    insta::assert_snapshot!(compile(config, modules));
}

#[test]
fn markdown_code_from_standalone_pages_is_not_trimmed() {
    let config = PackageConfig::default();
    let pages = vec![(
        "one",
        "
This is an example code snippet that should be indented
```gleam
pub fn indentation_test() {
  todo as \"This line should be indented by two spaces\"
}
```",
    )];
    insta::assert_snapshot!(compile_with_markdown_pages(config, vec![], pages));
}

#[test]
fn markdown_code_from_function_comment_is_trimmed() {
    let config = PackageConfig::default();
    let modules = vec![(
        "app.gleam",
        "
/// Here's an example code snippet:
/// ```
/// wibble
///   |> wobble
/// ```
///
pub fn indentation_test() {
  todo
}
",
    )];
    insta::assert_snapshot!(compile(config, modules));
}

#[test]
fn markdown_code_from_module_comment_is_trimmed() {
    let config = PackageConfig::default();
    let modules = vec![(
        "app.gleam",
        "
//// Here's an example code snippet:
//// ```
//// wibble
////   |> wobble
//// ```
////
",
    )];
    insta::assert_snapshot!(compile(config, modules));
}

#[test]
fn doc_for_commented_definitions_is_not_included_in_next_constant() {
    let config = PackageConfig::default();
    let modules = vec![(
        "app.gleam",
        "
/// Not included!
// pub fn wibble() {}

/// Included!
pub const wobble = 1
",
    )];
    assert!(!compile(config, modules).contains("Not included!"));
}

#[test]
fn doc_for_commented_definitions_is_not_included_in_next_type() {
    let config = PackageConfig::default();
    let modules = vec![(
        "app.gleam",
        "
/// Not included!
// pub fn wibble() {}

/// Included!
pub type Wibble {
  /// Wobble!
  Wobble
}
",
    )];
    assert!(!compile(config, modules).contains("Not included!"));
}

#[test]
fn doc_for_commented_definitions_is_not_included_in_next_function() {
    let config = PackageConfig::default();
    let modules = vec![(
        "app.gleam",
        "
/// Not included!
// pub fn wibble() {}

/// Included!
pub fn wobble(arg) {}
",
    )];
    assert!(!compile(config, modules).contains("Not included!"));
}

#[test]
fn doc_for_commented_definitions_is_not_included_in_next_type_alias() {
    let config = PackageConfig::default();
    let modules = vec![(
        "app.gleam",
        "
/// Not included!
// pub fn wibble() {}

/// Included!
pub type Wibble = Int
",
    )];
    assert!(!compile(config, modules).contains("Not included!"));
}

#[test]
fn source_link_for_github_repository() {
    let mut config = PackageConfig::default();
    config.repository = Repository::GitHub {
        user: "wibble".to_string(),
        repo: "wobble".to_string(),
        path: None,
    };

    let modules = vec![("app.gleam", "pub type Wibble = Int")];
    assert!(compile(config, modules)
        .contains("https://github.com/wibble/wobble/blob/v0.1.0/src/app.gleam#L1-L1"));
}

#[test]
fn source_link_for_github_repository_with_path() {
    let mut config = PackageConfig::default();
    config.repository = Repository::GitHub {
        user: "wibble".to_string(),
        repo: "wobble".to_string(),
        path: Some("path/to/package".to_string()),
    };

    let modules = vec![("app.gleam", "pub type Wibble = Int")];
    assert!(compile(config, modules).contains(
        "https://github.com/wibble/wobble/blob/v0.1.0/path/to/package/src/app.gleam#L1-L1"
    ));
}
