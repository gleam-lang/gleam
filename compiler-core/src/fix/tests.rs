use super::*;

fn fix(src: &str) -> String {
    parse_fix_and_format(&src.into(), Path::new("test")).unwrap()
}

#[test]
fn empty() {
    assert_eq!(fix(""), "\n")
}

#[test]
fn import_needed() {
    assert_eq!(
        fix("
pub fn main(y) {
  try x = y
  y
}
"),
        "import gleam/result

pub fn main(y) {
  use x <- result.then(y)
  y
}
"
    )
}

#[test]
fn import_not_needed() {
    assert_eq!(
        fix("import gleam/result

pub fn main(y) {
  try x = y
  y
}
"),
        "import gleam/result

pub fn main(y) {
  use x <- result.then(y)
  y
}
"
    )
}

#[test]
fn discarding_the_value() {
    assert_eq!(
        fix("import gleam/result

pub fn main(y) {
  try _ = y
  y
}
"),
        "import gleam/result

pub fn main(y) {
  use _ <- result.then(y)
  y
}
"
    )
}

#[test]
fn module_name_already_taken() {
    assert_eq!(
        fix("import one/result

pub fn main(y) {
  try _ = y
  y
}
"),
        "import gleam/result as gleam_result

import one/result

pub fn main(y) {
  use _ <- gleam_result.then(y)
  y
}
"
    )
}

#[test]
fn both_module_names_already_taken() {
    assert_eq!(
        fix("import two/result as gleam_result
import one/result

pub fn main(y) {
  try _ = y
  y
}
"),
        "import gleam/result as gleam_gleam_result

import two/result as gleam_result
import one/result

pub fn main(y) {
  use _ <- gleam_gleam_result.then(y)
  y
}
"
    )
}

#[test]
fn patterns() {
    assert_eq!(
        fix("
pub fn main(y) {
  try Box(x) = y
  x
}
"),
        "import gleam/result

pub fn main(y) {
  use Box(x) <- result.then(y)
  x
}
"
    )
}
