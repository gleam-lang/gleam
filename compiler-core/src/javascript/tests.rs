mod assignments;
mod blocks;
mod bools;
mod case_clause_guards;
mod custom_types;
mod externals;
mod functions;
mod lists;
mod modules;
mod numbers;
mod prelude;
mod strings;
mod try_;
mod tuples;

use pretty_assertions::assert_eq;

pub static CURRENT_PACKAGE: &str = "thepackage";

#[macro_export]
macro_rules! assert_js {
    ($src:expr, $erl:expr $(,)?) => {{
        use crate::javascript::*;
        let mut modules = std::collections::HashMap::new();
        let mut uid = 0;
        // DUPE: preludeinsertion
        // TODO: Currently we do this here and also in the tests. It would be better
        // to have one place where we create all this required state for use in each
        // place.
        let _ = modules.insert("gleam".to_string(), crate::type_::build_prelude(&mut uid));

        let (mut ast, _) = crate::parse::parse_module($src).expect("syntax error");
        ast.name = vec!["my".to_string(), "mod".to_string()];
        let ast = crate::type_::infer_module(
            &mut 0,
            ast,
            crate::build::Origin::Src,
            "thepackage",
            &modules,
            &mut vec![],
        )
        .expect("should successfully infer");
        let mut output = String::new();
        let line_numbers = LineNumbers::new($src);
        module(&ast, &line_numbers, &mut output).unwrap();
        assert_eq!(($src, output), ($src, $erl.to_string()));
    }};

    (($dep_package:expr, $dep_name:expr, $dep_src:expr), $src:expr, $erl:expr $(,)?) => {{
        use crate::javascript::*;
        let mut modules = std::collections::HashMap::new();
        let mut uid = 0;
        // DUPE: preludeinsertion
        // TODO: Currently we do this here and also in the tests. It would be better
        // to have one place where we create all this required state for use in each
        // place.
        let _ = modules.insert("gleam".to_string(), crate::type_::build_prelude(&mut uid));
        let (mut ast, _) = crate::parse::parse_module($dep_src).expect("dep syntax error");
        ast.name = $dep_name;
        let dep = crate::type_::infer_module(
            &mut 0,
            ast,
            crate::build::Origin::Src,
            $dep_package,
            &modules,
            &mut vec![],
        )
        .expect("should successfully infer");
        let _ = modules.insert($dep_name.join("/"), dep.type_info);
        let (mut ast, _) = crate::parse::parse_module($src).expect("syntax error");
        ast.name = vec!["my".to_string(), "mod".to_string()];
        let ast = crate::type_::infer_module(
            &mut 0,
            ast,
            crate::build::Origin::Src,
            CURRENT_PACKAGE,
            &modules,
            &mut vec![],
        )
        .expect("should successfully infer");
        let mut output = String::new();
        let line_numbers = LineNumbers::new($src);
        module(&ast, &line_numbers, &mut output).unwrap();
        assert_eq!(($src, output), ($src, $erl.to_string()));
    }};
}

#[test]
fn todo_throws_error() {
    assert_js!(
        r#"
fn go() {
    todo
}
"#,
        r#""use strict";

function go() {
  throw Object.assign(
    new Error("This has not yet been implemented"),
    { gleam_error: "todo", module: "my/mod", function: "go", line: 3 }
  )
}
"#
    );
    assert_js!(
        r#"
fn go() {
    todo("I should do this");
}
"#,
        r#""use strict";

function go() {
  throw Object.assign(
    new Error("I should do this"),
    { gleam_error: "todo", module: "my/mod", function: "go", line: 3 }
  )
}
"#
    );
}
