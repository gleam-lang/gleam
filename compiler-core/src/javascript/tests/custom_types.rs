use crate::assert_js;
use crate::javascript::tests::CURRENT_PACKAGE;

#[test]
fn zero_arity_literal() {
    assert_js!(
        r#"
type Mine {
    This
    ThatOneIsAMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchLongerVariant
}

fn go() {
    This
    ThatOneIsAMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchLongerVariant
}
"#,
        r#""use strict";

function go() {
  { type: "This" };
  return {
    type: "ThatOneIsAMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchLongerVariant"
  };
}
"#
    );
}

#[test]
fn zero_arity_const() {
    assert_js!(
        r#"
type Mine {
    This
    ThatOneIsAMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchLongerVariant
}

const this = This;
const that = ThatOneIsAMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchLongerVariant;
"#,
        r#""use strict";

const this$ = { type: "This" };

const that = {
  type: "ThatOneIsAMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchMuchLongerVariant"
};
"#
    );
}

#[test]
fn zero_arity_imported() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two }"#
        ),
        r#"import other
pub fn main() {
  other.Two
}"#,
        r#""use strict";

import * as Other from "../other.js";

export function main() {
  return { type: "Two" };
}
"#
    );
}

#[test]
fn zero_arity_imported_unqualified() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two }"#
        ),
        r#"import other.{Two}
pub fn main() {
  Two
}"#,
        r#""use strict";

import * as Other from "../other.js";

export function main() {
  return { type: "Two" };
}
"#
    );
}

// TODO
// #[test]
// fn zero_arity_imported_unqualified_aliased() {
//     assert_js!(
//         ( CURRENT_PACKAGE, vec!["other".to_string()], r#"pub type One { Two }"#),
//         r#"import other.{Two as Three}
// pub fn main() {
//   Three
// }"#,
//         r#""use strict";
//
// import * as Other from "../other.js";
// const { Two as Three } = other;
//
// export function main() {
//   return { type: "Two" };
// }
// "#
//     );
// }

#[test]
fn const_zero_arity_imported() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two }"#
        ),
        r#"import other
const x = other.Two
"#,
        r#""use strict";

const x = { type: "Two" };

import * as Other from "../other.js";
"#
    );
}

#[test]
fn const_zero_arity_imported_unqualified() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two }"#
        ),
        r#"import other.{Two}
const a = Two
"#,
        r#""use strict";

const a = { type: "Two" };

import * as Other from "../other.js";
"#
    );
}

// TODO
// #[test]
// fn const_zero_arity_imported_unqualified_aliased() {
//     assert_js!(
//         ( CURRENT_PACKAGE, vec!["other".to_string()], r#"pub type One { Two }"#),
//         r#"import other.{Two as Three}
// const a = Three
// "#,
//         r#""use strict";

// const a = { type: "Two" };

// import * as Other from "../other.js";
// const { Two as Three } = other;
// "#
//     );
// }

#[test]
fn const_with_fields() {
    assert_js!(
        r#"
type Mine {
  Mine(a: Int, b: Int)
}

const labels = Mine(b: 2, a: 1)
const no_labels = Mine(3, 4)
"#,
        r#""use strict";

const labels = { type: "Mine", a: 1, b: 2 };

const no_labels = { type: "Mine", a: 3, b: 4 };
"#
    );
}

#[test]
fn unnamed_fields() {
    assert_js!(
        r#"
type Ip{
    Ip(String)
}

const local = Ip("0.0.0.0")

fn build(x) {
    x("1.2.3.4")
}

fn go() {
    build(Ip)
    Ip("5.6.7.8")
}

fn destructure(x) {
  let Ip(raw) = x
  raw
}
"#,
        r#""use strict";

const local = { type: "Ip", 0: "0.0.0.0" };

function build(x) {
  return x("1.2.3.4");
}

function go() {
  build((var0) => { return { type: "Ip", 0: var0 }; });
  return { type: "Ip", 0: "5.6.7.8" };
}

function destructure(x) {
  if (x.type !== "Ip") throw new Error("Bad match");
  let raw = x[0];
  return raw;
}
"#
    );

    assert_js!(
        r#"
type TypeWithALongNameAndSeveralArguments{
  TypeWithALongNameAndSeveralArguments(String, String, String, String, String)
}


fn go() {
  TypeWithALongNameAndSeveralArguments
}
"#,
        r#""use strict";

function go() {
  return (var0, var1, var2, var3, var4) => {
    return {
      type: "TypeWithALongNameAndSeveralArguments",
      0: var0,
      1: var1,
      2: var2,
      3: var3,
      4: var4
    };
  };
}
"#
    );
}

#[test]
fn custom_type_with_named_fields() {
    assert_js!(
        r#"
type Cat {
  Cat(name: String, cuteness: Int)
}

const felix = Cat("Felix", 12)
const tom = Cat(cuteness: 1, name: "Tom")

fn go() {
  Cat("Nubi", 1)
  Cat(2, name: "Nubi")
  Cat(cuteness: 3, name: "Nubi")
}

fn update(cat) {
  Cat(..cat, name: "Sid")
  Cat(..cat, name: "Bartholemew Wonder Puss the Fourth !!!!!!!!!!!!!!!!")
}

fn access(cat: Cat) {
  cat.cuteness
}
"#,
        r#""use strict";

const felix = { type: "Cat", name: "Felix", cuteness: 12 };

const tom = { type: "Cat", name: "Tom", cuteness: 1 };

function go() {
  { type: "Cat", name: "Nubi", cuteness: 1 };
  { type: "Cat", name: "Nubi", cuteness: 2 };
  return { type: "Cat", name: "Nubi", cuteness: 3 };
}

function update(cat) {
  Object.assign({}, cat, { name: "Sid" });
  return Object.assign(
    {},
    cat,
    { name: "Bartholemew Wonder Puss the Fourth !!!!!!!!!!!!!!!!" }
  );
}

function access(cat) {
  return cat.cuteness;
}
"#
    );
}

#[test]
fn destructure_custom_type_with_named_fields() {
    assert_js!(
        r#"
type Cat {
  Cat(name: String, cuteness: Int)
}

fn go(cat) {
  let Cat(x, y) = cat
  let Cat(name: x, ..) = cat
  let Cat(cuteness: 4, name: x) = cat
  x
}

"#,
        r#""use strict";

function go(cat) {
  if (cat.type !== "Cat") throw new Error("Bad match");
  let x = cat.name;
  let y = cat.cuteness;
  if (cat.type !== "Cat") throw new Error("Bad match");
  let x$1 = cat.name;
  if (cat.type !== "Cat" || cat.cuteness !== 4) throw new Error("Bad match");
  let x$2 = cat.name;
  return x$2;
}
"#
    )
}

#[test]
fn nested_pattern_with_labels() {
    assert_js!(
        r#"pub type Box(x) { Box(a: Int, b: x) }
fn go(x) {
  case x {
    Box(a: _, b: Box(a: a, b: b)) -> a + b
    _ -> 1
  }
}
"#,
        r#""use strict";

function go(x) {
  if (x.type === "Box" && x.b.type === "Box") {
    let a = x.b.a;
    let b = x.b.b;
    return a + b;
  } else {
    return 1;
  }
}
"#
    );
}

#[test]
fn imported_no_label() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(Int) }"#
        ),
        r#"import other
pub fn main() {
  other.Two(1)
}"#,
        r#""use strict";

import * as Other from "../other.js";

export function main() {
  return { type: "Two", 0: 1 };
}
"#
    );
}

#[test]
fn imported_ignoring_label() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(field: Int) }"#
        ),
        r#"import other
pub fn main() {
  other.Two(1)
}"#,
        r#""use strict";

import * as Other from "../other.js";

export function main() {
  return { type: "Two", field: 1 };
}
"#
    );
}

#[test]
fn imported_using_label() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(field: Int) }"#
        ),
        r#"import other
pub fn main() {
  other.Two(field: 1)
}"#,
        r#""use strict";

import * as Other from "../other.js";

export function main() {
  return { type: "Two", field: 1 };
}
"#
    );
}

#[test]
fn imported_multiple_fields() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(a: Int, b: Int, c: Int) }"#
        ),
        r#"import other
pub fn main() {
  other.Two(b: 2, c: 3, a: 1)
}"#,
        r#""use strict";

import * as Other from "../other.js";

export function main() {
  return { type: "Two", a: 1, b: 2, c: 3 };
}
"#
    );
}

#[test]
fn unqualified_imported_no_label() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(Int) }"#
        ),
        r#"import other.{Two}
pub fn main() {
  Two(1)
}"#,
        r#""use strict";

import * as Other from "../other.js";

export function main() {
  return { type: "Two", 0: 1 };
}
"#
    );
}

#[test]
fn unqualified_imported_ignoring_label() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(field: Int) }"#
        ),
        r#"import other.{Two}
pub fn main() {
  Two(1)
}"#,
        r#""use strict";

import * as Other from "../other.js";

export function main() {
  return { type: "Two", field: 1 };
}
"#
    );
}

#[test]
fn unqualified_imported_using_label() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(field: Int) }"#
        ),
        r#"import other.{Two}
pub fn main() {
  Two(field: 1)
}"#,
        r#""use strict";

import * as Other from "../other.js";

export function main() {
  return { type: "Two", field: 1 };
}
"#
    );
}

#[test]
fn unqualified_imported_multiple_fields() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(a: Int, b: Int, c: Int) }"#
        ),
        r#"import other.{Two}
pub fn main() {
  Two(b: 2, c: 3, a: 1)
}"#,
        r#""use strict";

import * as Other from "../other.js";

export function main() {
  return { type: "Two", a: 1, b: 2, c: 3 };
}
"#
    );
}

#[test]
fn constructor_as_value() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(a: Int, b: Int, c: Int) }"#
        ),
        r#"import other
pub fn main() {
  other.Two
}"#,
        r#""use strict";

import * as Other from "../other.js";

export function main() {
  return (var0, var1, var2) => {
    return { type: "Two", a: var0, b: var1, c: var2 };
  };
}
"#
    );
}

#[test]
fn unqualified_constructor_as_value() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(a: Int, b: Int, c: Int) }"#
        ),
        r#"import other.{Two}
pub fn main() {
  Two
}"#,
        r#""use strict";

import * as Other from "../other.js";

export function main() {
  return (var0, var1, var2) => {
    return { type: "Two", a: var0, b: var1, c: var2 };
  };
}
"#
    );
}

#[test]
fn const_imported_no_label() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(Int) }"#
        ),
        r#"import other
pub const main = other.Two(1)
"#,
        r#""use strict";

export const main = { type: "Two", 0: 1 };

import * as Other from "../other.js";
"#
    );
}

#[test]
fn const_imported_ignoring_label() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(field: Int) }"#
        ),
        r#"import other
pub const main = other.Two(1)
"#,
        r#""use strict";

export const main = { type: "Two", field: 1 };

import * as Other from "../other.js";
"#
    );
}

#[test]
fn const_imported_using_label() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(field: Int) }"#
        ),
        r#"import other
pub const main = other.Two(field: 1)
"#,
        r#""use strict";

export const main = { type: "Two", field: 1 };

import * as Other from "../other.js";
"#
    );
}

#[test]
fn const_imported_multiple_fields() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(a: Int, b: Int, c: Int) }"#
        ),
        r#"import other
pub const main = other.Two(b: 2, c: 3, a: 1)
"#,
        r#""use strict";

export const main = { type: "Two", a: 1, b: 2, c: 3 };

import * as Other from "../other.js";
"#
    );
}

#[test]
fn const_unqualified_imported_no_label() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(Int) }"#
        ),
        r#"import other.{Two}
pub const main = Two(1)
"#,
        r#""use strict";

export const main = { type: "Two", 0: 1 };

import * as Other from "../other.js";
"#
    );
}

#[test]
fn const_unqualified_imported_ignoring_label() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(field: Int) }"#
        ),
        r#"import other.{Two}
pub const main = Two(1)
"#,
        r#""use strict";

export const main = { type: "Two", field: 1 };

import * as Other from "../other.js";
"#
    );
}

#[test]
fn const_unqualified_imported_using_label() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(field: Int) }"#
        ),
        r#"import other.{Two}
pub const main = Two(field: 1)
"#,
        r#""use strict";

export const main = { type: "Two", field: 1 };

import * as Other from "../other.js";
"#
    );
}

#[test]
fn const_unqualified_imported_multiple_fields() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(a: Int, b: Int, c: Int) }"#
        ),
        r#"import other.{Two}
pub const main = Two(b: 2, c: 3, a: 1)
"#,
        r#""use strict";

export const main = { type: "Two", a: 1, b: 2, c: 3 };

import * as Other from "../other.js";
"#
    );
}

#[test]
fn imported_pattern() {
    assert_js!(
        (
            CURRENT_PACKAGE,
            vec!["other".to_string()],
            r#"pub type One { Two(a: Int, b: Int, c: Int) }"#
        ),
        r#"import other.{Two}

pub fn main(x) {
  case x {
    Two(a: 1, ..) -> 1
    other.Two(b: 2, c: c, ..) -> c
    _ -> 3
  }
}
"#,
        r#""use strict";

import * as Other from "../other.js";

export function main(x) {
  if (x.type === "Two" && x.a === 1) {
    return 1;
  } else if (x.type === "Two" && x.b === 2) {
    let c = x.c;
    return c;
  } else {
    return 3;
  }
}
"#
    );
}
