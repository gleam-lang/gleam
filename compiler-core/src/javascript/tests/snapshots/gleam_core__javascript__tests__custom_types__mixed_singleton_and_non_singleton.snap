---
source: compiler-core/src/javascript/tests/custom_types.rs
expression: "\npub type Result {\n  Ok(value: Int)\n  Error\n}\n\npub fn is_error(r: Result) -> Bool {\n  r == Error\n}\n"
---
----- SOURCE CODE

pub type Result {
  Ok(value: Int)
  Error
}

pub fn is_error(r: Result) -> Bool {
  r == Error
}


----- COMPILED JAVASCRIPT
import { CustomType as $CustomType } from "../gleam.mjs";

export class Ok extends $CustomType {
  constructor(value) {
    super();
    this.value = value;
  }
}

export class Error extends $CustomType {}

export function is_error(r) {
  return r instanceof Error;
}
